fvm/src/kernel/default.rs

```diff
@@ -224,35 +224,32 @@ where
         t.record(Ok(self.get_self()?.map(|a| a.balance).unwrap_or_default()))
     }
 
-    fn self_destruct(&mut self, beneficiary: &Address) -> Result<()> {
+    fn self_destruct(&mut self) -> Result<()> {
         if self.read_only {
             return Err(syscall_error!(ReadOnly; "cannot self-destruct when read-only").into());
         }
 
-        // Idempotentcy: If the actor doesn't exist, this won't actually do anything. The current
+        // Idempotent: If the actor doesn't exist, this won't actually do anything. The current
         // balance will be zero, and `delete_actor_id` will be a no-op.
         let t = self
             .call_manager
             .charge_gas(self.call_manager.price_list().on_delete_actor())?;
 
+        // If there are remaining funds, burn them. We do this instead of letting the user to
+        // specify the beneficiary as:
```

Comment:
One thing I don't get is If there are remaining funds, why don't we just fail the self destruct altogether requiring the user to explicitly transfer (or burn) the funds himself?

Reply:
We considered that and I'm happy to go that way. I did it this way because it's most likely to "just work". E.g., you write a smart contract that never even expects to _have_ native tokens and someone sends you some anyways.

On the other hand, now that I revisit this... I'm more convinced we _should_ just fail. That ensures that fund transfers can only happen on "send" (and actor creation), not through other means.

Reply:
Well, kind of. On the other hand, I'm not sure if that matters. Basically, when an actor self destructs, it does kind of make sense to destroy the funds automatically. _Technically_ it's sending them to the burnt funds actor, but that's more of an implementation detail than anything.

Reply:
Achieving zero funds can be hard in cases where this account is paying for gas. While it might not be a worry right now (as key accounts cannot self-destruct), in case of Account Abstraction it gets harder.

Reply:
Yeah. In that world, I'll _require_ that the user pass specify that they want remaining funds burnt.